File: get-token.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"regexp"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/howeyc/gopass"
)

const BASE_URL = "https://www.meethue.com"

var httpClient http.Client

func init() {
	jar, err := cookiejar.New(nil)
	if err != nil {
		panic(err.Error())
	}
	httpClient = http.Client{Jar: jar}

}

func main() {
	username, password := getCredentials()
	token, err := getToken(username, password)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
	} else {
		fmt.Println("Token:", token)
	}
}

func getCredentials() (string, string) {
	var password string
	var username string
	flag.StringVar(&password, "password", "", "password")
	flag.StringVar(&username, "username", "", "username")
	flag.Parse()

	reader := bufio.NewReader(os.Stdin)
	if username == "" {
		fmt.Print("Username: ")
		line, _ := reader.ReadString('\n')
		username = strings.TrimSpace(line)
	}
	if password == "" {
		fmt.Print("Password: ")
		password = string(gopass.GetPasswdMasked())
	}

	return username, password
}

func getToken(username, password string) (string, error) {
	bridgeID, err := getBridgeID()
	if err != nil {
		return "", err
	}
	if err := startTokenSession(bridgeID); err != nil {
		return "", err
	}

	tokenURL, err := grantAccess(username, password)
	if err != nil {
		return "", err
	}

	return fetchToken(tokenURL)
}

func getBridgeID() (string, error) {
	url := BASE_URL + "/api/nupnp"
	resp, err := httpClient.Get(url)
	if err != nil {
		return "", fmt.Errorf("getBridgeID: request to %v failed: %v", url, err)
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("getBridgeID: reading response body failed", err)
	}

	var bridges []struct{ ID string }
	err = json.Unmarshal(body, &bridges)
	if err != nil {
		return "", fmt.Errorf("getBridgeID: parsing response body failed", err)
	}

	return bridges[0].ID, nil
}

func startTokenSession(bridgeID string) error {
	url := BASE_URL + "/en-us/api/gettoken?devicename=iPhone+5&appid=hueapp&deviceid=" + bridgeID
	resp, err := httpClient.Get(url)
	if err != nil {
		return fmt.Errorf("starTokenSession failed: %v", err)
	}
	defer resp.Body.Close()

	return nil
}

func grantAccess(email, password string) (string, error) {
	grantAccessURL := BASE_URL + "/en-us/api/getaccesstokengivepermission"
	form := url.Values{"email": {email}, "password": {password}}
	resp, err := httpClient.PostForm(grantAccessURL, form)
	if err != nil {
		return "", fmt.Errorf("grantAccess: error posting form to %v", grantAccessURL, err)
	}
	defer resp.Body.Close()

	gq, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return "", fmt.Errorf("grantAccess: Unable to parse response", err)
	}

	tokenURL, found := gq.Find("[data-role='yes']").Attr("href")
	if !found {
		return "", fmt.Errorf("Invalid credentials")
	} else {
		return tokenURL, nil
	}

}

func fetchToken(tokenURL string) (string, error) {
	fullTokenURL := BASE_URL + tokenURL
	resp, err := httpClient.Get(fullTokenURL)
	if err != nil {
		return "", fmt.Errorf("fetchToken: error getting url %v: %v", fullTokenURL, err)
	}
	defer resp.Body.Close()

	gq, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return "", fmt.Errorf("fetchToken: Unable to parse response", err)
	}
	successURL, found := gq.Find("a.button-primary").Attr("href")
	if !found {
		return "", fmt.Errorf("Unable to find token link in body")
	}

	tokenExtractRe := regexp.MustCompile("^phhueapp://sdk/login/(.+)$")
	matches := tokenExtractRe.FindStringSubmatch(successURL)
	if len(matches) == 0 {
		return "", fmt.Errorf("Unable to parse token from link")
	}

	return matches[1], nil
}

File: get-token.js
/*
  Get a remote access token to the Hue api.

  1. get device id. assumes in same lan
  2. go to login page, login to get cookie
  3. post grant url to activate token
  4. cookie should be in go back to app page
*/


import cheerio from 'cheerio';
import _ from 'lodash';
import request from 'request';
import minimist from 'minimist';
import readlineSync from 'readline-sync';

const BASE_URL = 'https://www.meethue.com';
const TOKEN_HREF_RE = /^phhueapp:\/\/sdk\/login\/(.+)$/;


function main() {
  return getCredentialFromCommandLine()
    .then(getToken)
    .then(printResults)
    .catch((err) => {
      console.log('ERR!', err.stack);
    })
  ;
}


function getCredentialFromCommandLine() {
  const argv = minimist(process.argv.slice(2));

  return Promise.resolve({
    email: promptIfNull('email', argv.email),
    password: promptIfNull('password', argv.password, {echo: false}),
  });
}

function promptIfNull(label, val, {echo} = {echo:true}) {
  if (val != null) {
    return val;
  } else {
    return readlineSync.question(`${label}: `, { hideEchoBack: !echo });
  }
}


export default function getToken(creds) {
  return getBridgeID()
    .then(startTokenSession)
    .then(() => grantAccess(creds))
    .then(fetchToken)
  ;
}

function getBridgeID() {
  const bridgeDiscoveryURL = `${BASE_URL}/api/nupnp`;
  return fetch(bridgeDiscoveryURL)
    .then(res => {
      const bridges = JSON.parse(res.body);
      const bridge = bridges[0];
      if (bridge == null || bridge.id == null) {
        throw new Error('bridge not found. Please ensure you are in the same network as your Hue Bridge.');
      } else {
        return bridge.id;
      }
    });
}

function startTokenSession(bridgeID) {
  const tokenSessionURL = `${BASE_URL}/en-us/api/gettoken?devicename=iPhone+5&appid=hueapp&deviceid=${bridgeID}`;
  return fetch(tokenSessionURL);
}

function grantAccess(creds) {
  const grantAccessURL = `${BASE_URL}/en-us/api/getaccesstokengivepermission`;
  const req = {
    method: 'POST',
    headers: { 'content-type': 'application/x-www-form-urlencoded' },
    form: creds,
  };
  return fetch(grantAccessURL, req).then((res) => {
    const $ = cheerio.load(res.body);
    const href = $('[data-role="yes"]').attr('href');
    if (!href) {
      throw new Error('Unable to grant access to bridge. Please verify your username and password');
    }
    return href;
  });
}

function fetchToken(tokenResultURL) {
  return fetch(BASE_URL + tokenResultURL).then((res) => {
    const $ = cheerio.load(res.body);
    const href = $('a.button-primary').attr('href');
    const token = TOKEN_HREF_RE.exec(href)[1];
    return token;
  });
}

function printResults(token) {
  console.log(token);
}

function fetch(url, settings) {
  const settingsWithJar =  _.merge({ jar: true }, settings);
  return new Promise((resolve, reject) => {
    request(url, settingsWithJar, (err, res) => {
      if (err != null) {
        reject(err);
      } else if (res.statusCode >= 400) {
        const httpErr = new Error(`${url}: unexpected status ${res.statusCode}. body: ${res.body}`);
        reject(httpErr);
      } else {
        resolve(res);
      }
    });
  });

}

if (require.main === module) {
  main();
}
